rfcs/2020-09-02-3684-metric-namespaces.md

```diff
@@ -0,0 +1,177 @@
+  namespace = "apache"
+
+[transforms.my_transform_id]
+  # General
+  type = "lua" # required
+  inputs = ["my_source_id"] # required
+  version = "2" # required
+
+  # Hooks
+  hooks.process = """
+  function (event, emit)
+    if event.metric.namespace == "apache" then
+      -- do something
+    end
+
+    emit(event)
+  end
+  """
+
+[sinks.prometheus]
+  type = "prometheus"
+  inputs = ["my_transform_id"]
+  address = "0.0.0.0:9598"
+  namespace = ""
+
+[sinks.cloudwatch]
+  type = "aws_cloudwatch_metrics"
+  inputs = ["my_transform_id"]
+  namespace = ""
+  region = "us-east-1"
+```
+
+Where the `prometheus` sink would simply output metrics with name prefixed by
+`apache_` and `aws_cloudwatch_metrics` would use it as the separate `Namespace`
+field in AWS API calls.
+
+Once [Make the `namespace` option on metrics sinks optional #3609](
+https://github.com/timberio/vector/issues/3609) is done. The sinks could look
+something like:
+
+```toml
+[sinks.my_sink_id]
+  type = "prometheus"
+  inputs = ["my_transform_id"]
+  address = "0.0.0.0:9598"
+  default_namespace = "unknown"
+```
+
+Where a namespace could be set for any metrics that do not already have one.
+
+## Rationale
+
+Currently, I don't think there is a way to tell if a metric already has a
+namespace to avoid setting an additional one in sinks the require it.
+
+## Prior Art
+
+- [telegraf](https://github.com/influxdata/telegraf/blob/b5fafb4c957b55701738f3ae78da4f54ffdec965/metric/metric.go#L12-L20).
+  They model it a bit differently with a `metric` having a number of fields
+  where the `name` is what we would call the `namespace` and `fields` is what we
+  would generate individual metrics for.
+
+## Drawbacks
+
+- Not all metric sources may have the concept of a "namespace" so we'll need to
+  figure out what do with it for those cases. It think prefixing it as
+  `prometheus` does would be a reasonable default.
+- This namespace concept may be confusing to users if none of their sinks or
+  sources use it
+
+## Alternatives
+
+### Model a metric as a set of measurements
+
+We could opt to model metrics closer to [how Telegraf does
+it](https://github.com/influxdata/telegraf/blob/b5fafb4c957b55701738f3ae78da4f54ffdec965/metric/metric.go#L12-L20)
+where we would encode all of the metrics for a given source as one metric with
+a set of fields.
+
+I didn't closely consider this option given that the proposed option seems
+reasonable and is a smaller change to the data model.
+
+## Outstanding Questions
+
+- Do we want to have the `prometheus` source parse the `namespace` out of
+  metrics it scrapes? The [naming
+  conventions](https://prometheus.io/docs/practices/naming/) suggest that all
+  metrics should start with one word describing the domain (or namespace)
+  followed by a `_` but there is requirement that prometheus endpoints satisfy
+  this. We could make it optional directive on the source to control parsing
+  metric namespaces.
```

Comment:
Would this affect backwards compatibility? With some transform configs I'm sure it would, but I wonder if it affects simpler pipelines.

Also probably worth noting that this would be a breaking change for at least transforms that operate on metric names.

Reply:
We could make this behavior opt-in for now to preserve compatibility with existing pipelines (and also to handle cases where the scraped endpoint isn't prefix a domain).

Reply:
I'm inclined to leave it for `remap` and just provide an example of doing it for the `prometheus` source.